Dependencies/UrlLib/Source/Unix/UrlRequest.cpp

```diff
@@ -76,6 +75,86 @@ namespace UrlLib
     private:
+        struct CurlMulti
+        {
+            CurlMulti()
+            : m_multiHandle(curl_multi_init())
+            {
+                assert(m_multiHandle != nullptr);
+                curl_multi_setopt(m_multiHandle, CURLMOPT_MAX_TOTAL_CONNECTIONS, 200);
+                curl_multi_setopt(m_multiHandle, CURLMOPT_MAX_HOST_CONNECTIONS, 6L);
+                m_thread = std::thread([this](){
+                    Loop();
+                });
+            }
+
+            ~CurlMulti()
+            {
+                m_cancelSource.cancel();
+                m_thread.join();
+                auto errCode = curl_multi_cleanup(m_multiHandle);
+                assert(errCode == CURLM_OK);
+            }
+
+            void AddHandle(CURL* handle)
+            {
+                std::lock_guard<std::mutex> guard(m_mutex);
+                CURLMcode errCode = curl_multi_add_handle(m_multiHandle, handle); 
+                assert(errCode == CURLM_OK);
+                (void)errCode;
+            }
+
+        private:
+            CURLM* m_multiHandle;
+            // can't add new handle when curl is inside a callback. To be certain, a mutex is used when performing calls.
+            std::mutex m_mutex;
+            arcana::cancellation_source m_cancelSource{};
+            std::thread m_thread;
+            void Loop()
+            {
+                while (!m_cancelSource.cancelled())
+                {
+                    int stillRunning;
+                    int numfds;
+                    int msgsLeft;
+                    if (curl_multi_wait(m_multiHandle, NULL, 0, 1000, &numfds) == CURLM_OK)
+                    {
+                        std::lock_guard<std::mutex> guard(m_mutex);
+                        if (curl_multi_perform(m_multiHandle, &stillRunning) != CURLM_OK)
+                        {
+                            throw std::runtime_error("CURL: Curl Multi perform error.");
+                        }
+                    }
+                    
+                    // See how the transfers went
+                    CURLMsg* m = nullptr;
+                    while ((m = curl_multi_info_read(m_multiHandle, &msgsLeft))) 
+                    {
+                        if (m->msg == CURLMSG_DONE) 
+                        {
+                            CURL* handle = m->easy_handle;
+                            UrlRequest::Impl *request;
+                            curl_easy_getinfo(handle, CURLINFO_PRIVATE, &request);
+                            if (m->data.result == CURLE_OK) 
+                            {
+                                long codep;
+                                curl_easy_getinfo(handle, CURLINFO_RESPONSE_CODE, &codep);
+
+                                // codep 0 for file access
+                                if (codep == 0 || codep == 200) 
+                                {
+                                    request->m_statusCode = UrlStatusCode::Ok;
+                                }
```

Comment:
What about failures? What happens if the url returns 404 or something else?

Reply:
I've open this issue : https://github.com/BabylonJS/BabylonNative/issues/427